In [8]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn import svm 

In [15]:
spam = pd.read_csv('/Users/user/Downloads/spam.csv', encoding='latin1')
print(np.array(spam))

[['ham'
  'Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'
  nan nan nan]
 ['ham' 'Ok lar... Joking wif u oni...' nan nan nan]
 ['spam'
  "Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's"
  nan nan nan]
 ...
 ['ham' 'Pity, * was in mood for that. So...any other suggestions?' nan
  nan nan]
 ['ham'
  "The guy did some bitching but I acted like i'd be interested in buying something else next week and he gave it to us for free"
  nan nan nan]
 ['ham' 'Rofl. Its true to its name' nan nan nan]]


In [16]:
print(spam.head)

<bound method NDFrame.head of         v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  

In [17]:
spam.drop(["Unnamed: 2", "Unnamed: 3", "Unnamed: 4"], axis=1, inplace=True)


In [18]:
print(spam)

        v1                                                 v2
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham              Will Ì_ b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]


In [30]:
spam.columns = ["Label", "Email"]
print(spam)

     Label                                              Email
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham              Will Ì_ b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]


In [38]:
from sklearn.metrics import classification_report, accuracy_score, precision_score, recall_score, f1_score, matthews_corrcoef, confusion_matrix


In [39]:
x = spam['Email']
y = spam["Label"]
x_train, x_test,y_train, y_test = train_test_split(x,y,test_size = 0.2)

In [41]:
cv = CountVectorizer()
features = cv.fit_transform(x_train)
print(features)

  (0, 3441)	1
  (0, 7709)	2
  (0, 3893)	1
  (0, 6397)	1
  (0, 6925)	2
  (0, 4367)	1
  (0, 7415)	1
  (0, 4129)	1
  (0, 3977)	1
  (0, 3610)	1
  (0, 7460)	1
  (0, 5918)	1
  (0, 7545)	2
  (0, 6810)	1
  (0, 2750)	1
  (0, 5715)	1
  (0, 6946)	1
  (0, 3017)	1
  (0, 7693)	1
  (0, 4986)	1
  (0, 4809)	1
  (1, 3893)	1
  (1, 6925)	1
  (1, 7415)	1
  (1, 3017)	1
  :	:
  (4455, 3456)	1
  (4456, 6925)	1
  (4456, 6810)	1
  (4456, 4986)	1
  (4456, 4809)	1
  (4456, 551)	1
  (4456, 4629)	1
  (4456, 1070)	1
  (4456, 4817)	1
  (4456, 7087)	1
  (4456, 5425)	1
  (4456, 5189)	1
  (4456, 312)	1
  (4456, 1927)	1
  (4456, 7587)	1
  (4456, 4052)	1
  (4456, 3280)	1
  (4456, 3749)	1
  (4456, 5247)	1
  (4456, 688)	1
  (4456, 4648)	1
  (4456, 500)	1
  (4456, 3585)	1
  (4456, 4070)	1
  (4456, 5016)	1


In [42]:
model = svm.SVC()
model.fit(features,y_train)

SVC()

In [43]:
features_test = cv.transform(x_test)
print("Accuracy: {}".format(model.score(features_test,y_test)))

Accuracy: 0.9856502242152466


In [ ]:
from sklearn.metrics import precision_score, recall_score

predictions = model.predict(features_test)

# Calculate precision
precision = precision_score(y_test, predictions, pos_label='spam')


# Calculate recall
recall = recall_score(y_test, predictions, pos_label='spam')

print("Precision: {}".format(precision))
print("Recall: {}".format(recall))


In summary, the Support Vector Classifier (SVC) model achieved an accuracy of approximately 98.57% on the test set, indicating its strong performance in classifying emails as spam or non-spam. Furthermore, the precision score for identifying 'spam' emails was approximately 99.19%, suggesting that the model correctly identified a high proportion of actual spam emails among all emails predicted as spam. However, the recall score for 'spam' emails was slightly lower at approximately 89.13%, indicating that the model missed identifying some spam emails present in the dataset. Nonetheless, the overall performance of the SVC model in terms of accuracy, precision, and recall indicates its effectiveness in identifying spam emails with high confidence while minimizing false positives.